### PR TITLE
CDK: emit `AirbyteTraceMessage` with exception trace information

### LINKED ISSUE #12472: Update the Python CDK to emit `AirbyteTraceMessage`
Full context in parent issue https://github.com/airbytehq/airbyte/issues/12322

The python cdk should emit `AirbyteTraceMessage` with error information when something goes wrong.

### PR DESCRIPTION
**Pull request by pedroslopez**
## What
`AirbyteTraceMessage` has been introduced in the protocol to allow better debugging and attribution of failures when something in the connector goes wrong. 

This change starts emitting `AirbyteTraceMessage`s generically whenever an exception is raised. A future PR will make it easier to return better user-friendly messages.

closes #12472 

## How
- Use `sys.excepthook` to register a global handler for uncaught exceptions. This handler is in charge of emitting the generic `AirbyteTraceMessage`s.
- Introduce an `AirbyteTracedException` class, which can be used to customize the trace message and provide better error messages.
- Refactor the logger's secret filtering logic out into an `AirbyteSecretHelper` so it can be re-used to filter out secrets from trace messages.

## Recommended reading order
1. `utils/traced_exception.py`
2. `exception_handler.py`
3. tests

### CODE REVIEW

File: airbyte-cdk/python/airbyte_cdk/sources/utils/transform.py
Comment (pedroslopez) [inline airbyte-cdk/python/airbyte_cdk/sources/utils/transform.py:12]: A few places were still using the old, deprecated `AirbyteLogger` - aside from deprecation warnings, this was causing some circular dependency issues when I refactored the `AirbyteSecretHelper` out from `logger.py` into `airbyte_secrets_utils.py`.

File: airbyte-cdk/python/airbyte_cdk/utils/airbyte_secrets_utils.py
Comment (pedroslopez) [inline airbyte-cdk/python/airbyte_cdk/utils/airbyte_secrets_utils.py:36]: Moved out the logic from `AirbyteLogFormatter` so it can be re-used (using this for trace messages now as well)

File: airbyte-cdk/python/airbyte_cdk/utils/traced_exception.py
Comment (pedroslopez) [inline airbyte-cdk/python/airbyte_cdk/utils/traced_exception.py:40]: any suggestions on what this generic default message should be are welcome 😄

File: airbyte-cdk/python/airbyte_cdk/exception_handler.py
Comment (pedroslopez) [inline airbyte-cdk/python/airbyte_cdk/exception_handler.py:30]: This is the main part that is in charge of emitting the `AirbyteTraceMessage`: If the exception we get is already an `AirbyteTracedException` (which can be thrown from anywhere with appropriate messages and fields) we use it, if not then we build one from the exception. This is then turned into an `AirbyteTraceMessage` and filtered for secret values.

File: airbyte-cdk/python/airbyte_cdk/utils/airbyte_secrets_utils.py
Comment (Phlair) [inline airbyte-cdk/python/airbyte_cdk/utils/airbyte_secrets_utils.py:23]: Arguably this is not very pythonic in that we're defining a class with the intention to never instantiate it, and using classmethods to update and use a class attribute. This sort of global attribute storage / updating / usage could be done at the module level, e.g. at top level of this .py
```
__SECRETS_FROM_CONFIG = []

def update_secrets(secrets: List[str]):
        global __SECRETS_FROM_CONFIG
        __SECRETS_FROM_CONFIG = secrets

def filter_secrets(string: str) -> str:
        global __SECRETS_FROM_CONFIG
        for secret in __SECRETS_FROM_CONFIG:
            string = string.replace(secret, "****")
        return string
```
then elsewhere to use
```
from airbyte_secret_utils import update_secrets, filter_secrets
# and use the functions directly
```

This achieves exactly the same thing but more pythonic imo.

File: airbyte-cdk/python/airbyte_cdk/utils/traced_exception.py
Comment (Phlair) [reply airbyte-cdk/python/airbyte_cdk/utils/traced_exception.py:40]: this isn't necessarily source, could also be destination.

How about `Something went wrong in the connector. See the logs for more details.` ?
Pointing them at the logs seems sensible given that this is always going to be non-specific at this level.

**Review by Phlair — COMMENTED** [review]
Awesome! Mostly just comments on being pythonic, one open question around the `AirbyteLogger` which should be resolved before approve I think.

File: airbyte-cdk/python/airbyte_cdk/utils/traced_exception.py
Comment (Phlair) [inline airbyte-cdk/python/airbyte_cdk/utils/traced_exception.py:31]: this is a Double in the protocol so should be a `float` here in python

Comment (pedroslopez) [reply airbyte-cdk/python/airbyte_cdk/utils/traced_exception.py:31]: Oh interesting - the `RecordMessage`'s `emitted_at` is an `int` 🤔  good catch!

File: airbyte-cdk/python/airbyte_cdk/sources/utils/transform.py
Comment (Phlair) [reply airbyte-cdk/python/airbyte_cdk/sources/utils/transform.py:12]: clarified on slack, I misunderstood, we're all good here
i.e. 
> ~~ensure~~ we aren't breaking AirbyteLogger with this change

File: airbyte-cdk/python/airbyte_cdk/utils/airbyte_secrets_utils.py
Comment (pedroslopez) [reply airbyte-cdk/python/airbyte_cdk/utils/airbyte_secrets_utils.py]: Interesting! Thanks for the tip, changed in [94956f2](https://github.com/airbytehq/airbyte/pull/12593/commits/94956f25e592110f1ec48a5b84f3414d4170f963)

File: airbyte-cdk/python/airbyte_cdk/utils/traced_exception.py
Comment (evantahler) [reply airbyte-cdk/python/airbyte_cdk/utils/traced_exception.py]: Are there any environment variables or system arguments we can use to know if we are a source or destination?  e.g. looking at the docker container name (source-mailchimp)?

File: airbyte-cdk/python/airbyte_cdk/exception_handler.py
Comment (evantahler) [inline airbyte-cdk/python/airbyte_cdk/exception_handler.py:36]: This is cool - like Node's `process.on('uncaughtException')`

File: airbyte-cdk/python/airbyte_cdk/exception_handler.py
Comment (evantahler) [inline airbyte-cdk/python/airbyte_cdk/exception_handler.py:34]: [Ergonomic]: maybe the utility to emit the `AirbyteTracedException` as JSON should be built-in:

```python
traced_exc = (
            exception_value
            if issubclass(exception_type, AirbyteTracedException)
            else AirbyteTracedException.from_exception(exception_value)
        ).emit() # Where emit() gets the message and prints as JSON
```

File: airbyte-cdk/python/unit_tests/test_exception_handler.py
Comment (evantahler) [inline airbyte-cdk/python/unit_tests/test_exception_handler.py:42]: That's a neat way to test that a process crashed!

**Review by evantahler — APPROVED** [review]
I defer to other's Pythonic notes, but 👍 from me!  The ergonomic suggestion is non-blocking.

**Review by Phlair — APPROVED** [review]
LGTM! Nice one 👍

File: airbyte-cdk/python/airbyte_cdk/exception_handler.py
Comment (pedroslopez) [reply airbyte-cdk/python/airbyte_cdk/exception_handler.py]: Nice, done in [313b445](https://github.com/airbytehq/airbyte/pull/12593/commits/313b445cb4ab95f40c56aea52cbb68388f0580d8). 👍🏼

**Pull request merged**
